lib/iris/experimental/ugrid.py

```diff
@@ -998,97 +1030,614 @@ def equal(self, other, lenient=None):
+        edge=None,
+    ):
+        return self._remove(
+            item=item,
+            standard_name=standard_name,
+            long_name=long_name,
+            var_name=var_name,
+            attributes=attributes,
+            axis=axis,
+            node=node,
+            edge=edge,
+            asdict=True,
+        )
+
+
+class _Mesh2DCoordinateManager(_Mesh1DCoordinateManager):
+    OPTIONAL = (
+        "edge_x",
+        "edge_y",
+        "face_x",
+        "face_y",
+    )
+
+    def __init__(
+        self,
+        node_x,
+        node_y,
+        edge_x=None,
+        edge_y=None,
+        face_x=None,
+        face_y=None,
+    ):
+        super().__init__(node_x, node_y, edge_x=edge_x, edge_y=edge_y)
+
+        # optional coordinates
+        self.face_x = face_x
+        self.face_y = face_y
+
+    def __getstate__(self):
+        # TBD
+        pass
+
+    def __setstate__(self, state):
+        pass
+
+    @property
+    def _face_shape(self):
+        return self._shape(location="face")
+
+    @property
+    def all_coords(self):
+        return Mesh2DCoords(**self._members)
+
+    @property
+    def face_coords(self):
+        return MeshFaceCoords(face_x=self.face_x, face_y=self.face_y)
+
+    @property
+    def face_x(self):
+        return self._members["face_x"]
+
+    @face_x.setter
+    def face_x(self, coord):
+        self._setter(
+            location="face", axis="x", coord=coord, shape=self._face_shape
+        )
+
+    @property
+    def face_y(self):
+        return self._members["face_y"]
+
+    @face_y.setter
+    def face_y(self, coord):
+        self._setter(
+            location="face", axis="y", coord=coord, shape=self._face_shape
+        )
+
+    def add(
+        self,
+        node_x=None,
+        node_y=None,
+        edge_x=None,
+        edge_y=None,
+        face_x=None,
+        face_y=None,
+    ):
+        super().add(node_x=node_x, node_y=node_y, edge_x=edge_x, edge_y=edge_y)
+
+        # deal with the optional members
+        if face_x is not None:
+            self.face_x = face_x
+        if face_y is not None:
+            self.face_y = face_y
+
+    def coord(
+        self,
+        item=None,
+        standard_name=None,
+        long_name=None,
+        var_name=None,
+        attributes=None,
+        axis=None,
+        node=None,
+        edge=None,
+        face=None,
+        asdict=True,
+    ):
+        return self._coord(
+            item=item,
+            standard_name=standard_name,
+            long_name=long_name,
+            var_name=var_name,
+            attributes=attributes,
+            axis=axis,
+            node=node,
+            edge=edge,
+            face=face,
+            asdict=asdict,
+        )
+
+    def coords(
+        self,
+        item=None,
+        standard_name=None,
+        long_name=None,
+        var_name=None,
+        attributes=None,
+        axis=None,
+        node=None,
+        edge=None,
+        face=None,
```

Comment:
The concept of allowing / not allowing the `face` parameter depending on the subclass is very explicit and helpful, but it does require a **lot** of code, most of which is repeated, and I don't think the benefit is worth the cost.

If one passes `face=True`, but face coordinates can't exist, it's still perfectly correct to not return any coordinates. This would mean you wouldn't need to override the method, nor to build it on top of a private method - that's a big reduction in complexity, and the disadvantage is minimal given the following:

* We can make this possibility clear in the docstring.
* Anyone interacting with the `Mesh` won't see the distinction anyway, since they get a search method that's agnostic to the dimensionality.
* If necessary we could log a warning at the `Mesh` level if the code detects `self.topology_dimension` is too low to allow `face`s (or `volume`s).

Reply:
Cool. Understand this, let me thing about it and we can align either way

Reply:
@trexfeathers Okay, you've convinced me... let's roll with that and see how it plays out.

However, I'm against adopting this strategy for `add` and `remove`, as this means that the caller really doesn't understand their underlying data i.e., you're trying to add or remove an face on a 1D geometric element, which really is complete nonsense.

Totally happy to relax this for filtering though